### PR TITLE
rbd/mirror: set replayer status as resync invalid if no remote primary available

### PR DESCRIPTION
**Pull request by ideepika**
this covers the case where the remote is non-primary and we try resync
from it, deleting our local image and having no knowing primary image to
sync from.
Mark such resync as invalid and update replayer status.

fixes: https://tracker.ceph.com/issues/54098

```
a1:
  global_id:   2c25ef7b-448d-451b-80a6-9cc136729c1e
  state:       up+unknown
  description: remote image demoted
  service:     admin on vossi03
  last_update: 2022-02-09 15:41:09
  peer_sites:
    name: secondary
    state: up+stopped
    description: resync invalid, remote image is non-primary
    last_update: 2022-02-09 15:41:09
```

Signed-off-by: Deepika Upadhyay <dupadhya@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by trociny — APPROVED** [review]
LGTM.

I am slightly concerned with using `-EPERM` here, because theoretically this error might be return by librbd or ceph due to completely different reason and this may confuse. I was thinking to just return -EREMOTEIO so the status description would be "remote image is non-primary".

But your description looks more useful/informative, and actually my concern about `-EPERM` is very week, so I am fine with the current version. Just wanted to mention it just in case.

And I would combine the first and the third (mockup test fixup) commits, because they are related (it is difficult to understand the necessity of the mockup test change without knowing the rbd-mirror change).

File: qa/workunits/rbd/rbd_mirror_snapshot.sh [hunk 1]
```diff
@@ -440,6 +440,16 @@ wait_for_image_replay_started ${CLUSTER1} ${POOL} ${image}
 wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${image} 'up+replaying'
 compare_images ${POOL} ${image}
 
+testlog "TEST: request image resync while no known primary exists"
+demote_image ${CLUSTER1} ${POOL} ${image}
+request_resync_image ${CLUSTER1} ${POOL} ${image} image_id
+wait_for_image_present ${CLUSTER1} ${POOL} ${image} 'present' ${image_id}
+wait_for_image_present ${CLUSTER1} ${POOL} ${image} 'present'
+wait_for_image_replay_started ${CLUSTER1} ${POOL} ${image}
+wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${image} 'up+stopped'
```
Comment (MrFreezeex) [inline qa/workunits/rbd/rbd_mirror_snapshot.sh:449]: nit: you could add your new status message in this test as well
`wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${image} 'up+stopped' 'resync invalid, remote image is non-primary'`

**Review by MrFreezeex — APPROVED** [review]
LGTM

Maybe we could have a similar test in the journal workunit though?

**Comment by trociny** [comment]
+1 for the test in journal too. 

Ideally in some future we would want to combine the journal and snapshot test scripts into one generic, that would test either journal or snapshot depending on a value of an evn variable, to avoid the code duplication. But this is definitely not for this PR.

**Review by idryomov — CHANGES_REQUESTED** [review]
This PR is superseded by https://github.com/ceph/ceph/pull/46759 which is broader change addressing the actual root cause.  I only added unit tests there though, so continuing on with adding an integration test here might be worthwhile.

Unfortunately, no distinctive status would be reported as the replayer is shut down before it gets a chance to check whether a resync is requested.

**Pull request closed without merging**